src/slash-bedrock/share/brl-fetch/distros/exherbo

```diff
@@ -0,0 +1,147 @@
+#!/bedrock/libexec/busybox sh
+#
+# Exherbo Linux bootstrap support
+# Copyright 2019 Wulf C. Krueger <philantrop@exherbo.org>
+# Distributed under the terms of the GNU General Public License v2
+#
+# based in part upon "Gentoo Linux bootstrap support" which is:
+#      This program is free software; you can redistribute it and/or
+#      modify it under the terms of the GNU General Public License
+#      version 2 as published by the Free Software Foundation.
+#
+#	Copyright (c) 2016-2019 Daniel Thau <danthau@bedrocklinux.org>
+
+. /bedrock/share/common-code
+trap 'fetch_abort "Unexpected error occurred."' EXIT
+
+check_supported() {
+	true
+}
+
+speed_test_url() {
+	echo "/stages/sha1sum"
+}
+
+list_mirrors() {
+	echo "https://dev.exherbo.org/"
+}
+
+brl_arch_to_distro() {
+	case "${1}" in
+	     "x86_64") echo "amd64" ;;
+	     *) abort "brl does not know how to translate arch \"${1}\" to ${distro:-} format" ;;
+	esac
+}
+
+list_architectures() {
+	cat <<EOF
+x86_64
+EOF
+}
+
+default_release() {
+	echo "rolling"
+}
+
+list_releases() {
+	echo "rolling"
+}
+
+fetch() {
+        paludis_builddir=/var/paludis
+        paludis_repo_dir=/etc/paludis/repositories
+        paludis_user=paludisbuild
+        paludis_group=paludisbuild
+
+	ex_download=/var/tmp/brl-fetch
+	ex_stage=exherbo-x86_64-pc-linux-gnu-current.tar.xz
+	ex_checksum=sha1sum
+
+	mkdir -p "${ex_download}"
+
+	step "Downloading bootstrap software"
+	(
+	    if [[ -f ${ex_download}/${ex_stage} ]]; then
+	        cp "${ex_download}/${ex_stage}" "${bootstrap_dir:-}"
+		cp "${ex_download}/${ex_checksum}" "${bootstrap_dir:-}"
+	    else
+		wget ${target_mirror:-}/stages/"${ex_stage}" -O "${bootstrap_dir:-}"/"${ex_stage}"
+		wget ${target_mirror:-}/stages/"${ex_checksum}" -O "${bootstrap_dir:-}"/"${ex_checksum}"
+	    fi
```

Comment:
Yes, that would be useful as a global feature. 

I'd like to ask for an exception here, though, since we simply don't have mirrors for our stages. Everyone downloading a stage will always hit our main development machine.
I'll gladly replace it with a well-designed alternative but for the moment I'd appreciate it if we could leave it in.

Reply:
Ah, I was modelling this for debugging/injecting, not caching.  Minimizing mirror usage is completely reasonable.  I *really* don't want Bedrock getting a bad reputation for abusing mirrors.

I'd still prefer a some discussion on how to implement it with thought towards a generalized `brl fetch` solution before merging any `brl fetch` back end with such a feature.  Instead of removing caching from this MR, we can just work on the generalized solution first, provided you don't mind waiting a bit longer for this to merge.

Here's how I'm imagining a generalized `brl fetch` caching system right now:

- `/bedrock/libexec/brl-fetch` sets some variable - I'm currently thinking `${cache}` but am up for suggestion - that each back end inherits and uses.
- `${cache}` expands to `/bedrock/var/cache/brl-fetch/${distro}-${target_release}-${target_arch}`.  While each back-end *could* handle release/arch concerns, I don't think it's worth the complexity; easier to just have `brl fetch` segregate them.
- `brl fetch` will flock the cache directory.  If another `brl fetch` instance is grabbing the same release of the same distro for the same architecture at the same time, the new one errors out.  While it should definitely be possible to handle multiple simultaneous instances, I don't think it's worth the complexity.  Moreover, with caching, sequential access should end up being faster.
- `bedrock.conf` will contain some setting for how long the `brl fetch` cache should be kept around.  Calls to `brl fetch` after this time period will clear the cache.  Off the cuff I'm thinking of calling the section `[brl-fetch-cache]`, the key `expiration`, adding a comment explaining the unit is in days, defaulting this to `90`, but I'm definitely up for suggestion.
- The cache directory's timestamp is updated every `brl fetch` for the given distro/release/arch.  If the cache doesn't change, it sticks around.  This could be useful for distros like Debian that could re-use the same cached value for potentially years.
- Back ends are expected to check if the cached value is out of date every `brl fetch`.  If there are things like checksums, they should be re-downloaded every `brl fetch`.

I have not spent a lot of time thinking this through.  Feel free to push back on any detail, raise counter proposals, etc.

If this ends up being the system, I can probably add the `bedrock.conf` setting and shared caching code to `brl fetch` reasonably quickly.  May take me a bit to implement for all the other back ends, but they don't have to all get it in the same release.  I think the only changes you would need here are:

- Using the inherited `${cache}` (or rather `${cache:-}` to make `make check` happy, as it does not know the variable will be inherited).
- Changing the code to download the checksum every time, clear the old cached value if it is out of date or corrupt, then (re)download the new item.

Reply:
Sounds good to me.